src/datahike/db.cljc

```diff
@@ -1149,19 +1149,72 @@
       (update-in [:db-after] with-datom datom)
       (update-in [:tx-data] conj datom)))
 
+
+
+(defn validate-datom-upsert [db ^Datom datom]
+  (when (is-attr? db (.-a datom) :db/unique)
+    (when-let [old (first (-datoms db :avet [(.-a datom) (.-v datom)]))]
+      (when-not (= (.-e datom) (.-e old))
+        (raise "Cannot add " datom " because of unique constraint: " old
+          {:error     :transact/unique
+           :attribute (.-a datom)
+           :datom     datom})))))
+
+
+(defn- with-datom-upsert [db ^Datom datom]
+  (validate-datom-upsert db datom)
+  (let [indexing?      (indexing? db (.-a datom))
+        schema?        (ds/schema-attr? (.-a datom))
+        keep-history?  (and (-keep-history? db) (not (no-history? db (.-a datom))))
+        db             (if (or keep-history? indexing?)
```

Comment:
This means that when we want to keep the history or index we still use `-search` and B+-tree lookups here. Have you thought about what would be necessary to avoid that as well? This is necessary because we need `v` and `t` for the Datom, right? Maybe another type of UpsertOp could extract this once it arrives in the leaf and only insert itself (not deleting `old`)?